File: python/examples/websocket_client.py
"""WebSocket Client - Real-time event streaming."""

import asyncio, json, os
import websockets
from dotenv import load_dotenv

load_dotenv()

WS_URL = os.getenv("PROPR_WS_URL", "wss://api.propr.xyz/ws")
API_KEY = os.getenv("PROPR_API_KEY")

async def listen():
    async with websockets.connect(
        WS_URL,
        additional_headers={"X-API-Key": API_KEY},
        ping_interval=20, ping_timeout=10,
    ) as ws:
        async for raw in ws:
            msg = json.loads(raw)
            event_type = msg.get("type")
            data = msg.get("data", msg)

            if event_type == "position.updated":
                print(f"Position: {data.get('base')} uPnL={data.get('unrealizedPnl')}")
            elif event_type == "order.filled":
                print(f"Filled: {data.get('side')} {data.get('base')} @ {data.get('averageFillPrice')}")
            elif event_type == "trade.created":
                print(f"Trade: {data.get('type')} {data.get('quantity')} {data.get('base')}")

asyncio.run(listen())

File: javascript/examples/websocket-client.ts
/**
 * WebSocket Client - Real-time event streaming.
 *
 * npm install ws
 */

import WebSocket from 'ws';

const WS_URL = process.env.PROPR_WS_URL || 'wss://api.propr.xyz/ws';
const API_KEY = process.env.PROPR_API_KEY;

function connect() {
  const ws = new WebSocket(WS_URL, {
    headers: { 'X-API-Key': API_KEY! },
  });

  ws.on('open', () => console.log('Connected'));

  ws.on('message', (raw) => {
    const msg = JSON.parse(raw.toString());
    const { type, data } = msg;

    switch (type) {
      case 'connected':
        console.log(`Authenticated: ${data.userId}`);
        break;
      case 'position.updated':
        console.log(`Position: ${data.base} uPnL=${data.unrealizedPnl}`);
        break;
      case 'order.filled':
        console.log(`Filled: ${data.side} ${data.base} @ ${data.averageFillPrice}`);
        break;
      case 'trade.created':
        console.log(`Trade: ${data.type} ${data.quantity} ${data.base}`);
        break;
      default:
        console.log(`Event: ${type}`);
    }
  });

  ws.on('close', () => {
    console.log('Disconnected, reconnecting in 5s...');
    setTimeout(connect, 5000);
  });

  ws.on('error', (err) => console.error('WS error:', err.message));
}

connect();
